pkg/plugins/policies/meshtimeout/plugin/xds/configurer.go

```diff
@@ -0,0 +1,113 @@
+	defaultStreamIdleTimeout     = "30m"
+	defaultMaxStreamDuration     = "0s"
+	defaultMaxConnectionDuration = "0s"
+)
+
+type Configurer struct {
+	Conf     api.Conf
+	Protocol core_mesh.Protocol
+}
+
+func (c *Configurer) ConfigureListener(filterChain *envoy_listener.FilterChain) error {
+	httpTimeouts := func(hcm *envoy_hcm.HttpConnectionManager) error {
+		if c.Conf.Http != nil {
+			hcm.StreamIdleTimeout = toProtoDurationOrDefault(c.Conf.Http.StreamIdleTimeout, defaultStreamIdleTimeout)
+			configureRequestTimeout(hcm.GetRouteConfig(), toProtoDurationOrDefault(c.Conf.Http.RequestTimeout, defaultRequestTimeout))
+		} else {
+			hcm.StreamIdleTimeout = defaultToProtoDuration(defaultStreamIdleTimeout)
+			configureRequestTimeout(hcm.GetRouteConfig(), defaultToProtoDuration(defaultRequestTimeout))
+		}
+		return nil
+	}
+	tcpTimeouts := func(proxy *envoy_tcp.TcpProxy) error {
+		proxy.IdleTimeout = toProtoDurationOrDefault(c.Conf.IdleTimeout, defaultIdleTimeout)
+		return nil
+	}
+	switch c.Protocol {
+	case core_mesh.ProtocolHTTP, core_mesh.ProtocolHTTP2:
+		if err := listeners_v3.UpdateHTTPConnectionManager(filterChain, httpTimeouts); err != nil && !errors.Is(err, &listeners_v3.UnexpectedFilterConfigTypeError{}) {
+			return err
+		}
+	case core_mesh.ProtocolUnknown, core_mesh.ProtocolTCP, core_mesh.ProtocolKafka:
+		if err := listeners_v3.UpdateTCPProxy(filterChain, tcpTimeouts); err != nil && !errors.Is(err, &listeners_v3.UnexpectedFilterConfigTypeError{}) {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *Configurer) ConfigureCluster(cluster *envoy_cluster.Cluster) error {
+	cluster.ConnectTimeout = toProtoDurationOrDefault(c.Conf.ConnectionTimeout, defaultConnectionTimeout)
+	switch c.Protocol {
+	case core_mesh.ProtocolHTTP, core_mesh.ProtocolHTTP2:
+		err := clusters.UpdateCommonHttpProtocolOptions(cluster, func(options *envoy_upstream_http.HttpProtocolOptions) {
+			if options.CommonHttpProtocolOptions == nil {
+				options.CommonHttpProtocolOptions = &envoy_core.HttpProtocolOptions{}
+			}
+			commonHttp := options.CommonHttpProtocolOptions
+			commonHttp.IdleTimeout = toProtoDurationOrDefault(c.Conf.IdleTimeout, defaultIdleTimeout)
+			if c.Conf.Http != nil {
+				commonHttp.MaxStreamDuration = toProtoDurationOrDefault(c.Conf.Http.MaxStreamDuration, defaultMaxStreamDuration)
+				commonHttp.MaxConnectionDuration = toProtoDurationOrDefault(c.Conf.Http.MaxConnectionDuration, defaultMaxConnectionDuration)
+			} else {
+				commonHttp.MaxStreamDuration = defaultToProtoDuration(defaultMaxStreamDuration)
+				commonHttp.MaxConnectionDuration = defaultToProtoDuration(defaultMaxConnectionDuration)
+			}
+		})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func configureRequestTimeout(routeConfiguration *envoy_route.RouteConfiguration, timeout *durationpb.Duration) {
+	for _, vh := range routeConfiguration.VirtualHosts {
+		for _, route := range vh.Routes {
+			routeAction := route.GetRoute()
+			if routeAction != nil {
```

Comment:
I am not sure if this covers all the cases. I am working on a similar thing and I think when you define the source for which there is no route, we probably should create one. This implementation only edits the current route but does not create new ones. So in case when we want to create a timeout only for `service-a` we won't get it.

Reply:
I think that this approach is enough for timeouts. But question is how this will behave when we will combine multiple policies that create routes in the meantime?

Reply:
Yes, that might be problematic, when another plugin creates a route and Timeout is applied before this the new route won't get this configuration.